lib/ar_transaction_changes.rb

```diff
@@ -36,12 +36,31 @@ def write_attribute(attr_name, value)
 
   def _store_transaction_changed_attributes(attr_name)
     attr_name = attr_name.to_s
-    old_value = read_attribute(attr_name)
+    old_value = _read_attribute_for_transaction(attr_name)
     ret = yield
-    new_value = read_attribute(attr_name)
+    new_value = _read_attribute_for_transaction(attr_name)
     unless transaction_changed_attributes.key?(attr_name) || new_value == old_value
-      transaction_changed_attributes[attr_name] = old_value
+      attribute = @attributes[attr_name]
+      transaction_changed_attributes[attr_name] = if attribute.type.is_a?(::ActiveRecord::Type::Serialized)
+        attribute.type.deserialize(old_value)
+      else
+        old_value
+      end
+      transaction_changed_attributes
     end
     ret
   end
+
+  def _read_attribute_for_transaction(attr_name)
+    attribute = @attributes[attr_name]
+    if attribute.type.is_a?(::ActiveRecord::Type::Serialized)
```

Comment:
Are serialized attributes actually a special case?  It seems like the same problem can be seen with string attributes, which also seem to get duplicated on reads.

For example, before the regression was introduced to this library, the following modified test passes

```diff
   def test_transaction_changes_for_update
-    @user.name = "Dillon"
+    name = "Dillon"
+    @user.name = name
+    name.upcase!
     @user.save!
-    assert_equal ["Dylan", "Dillon"], @user.stored_transaction_changes["name"]
+    assert_equal ["Dylan", "DILLON"], @user.stored_transaction_changes["name"]
   end
```

The same test against this branch results in a failure

```
  1) Failure:
TransactionChangesTest#test_transaction_changes_for_update [test/transaction_changes_test.rb:25]:
Expected: ["Dylan", "DILLON"]
  Actual: ["Dylan", "Dillon"]
```

due to the same regression.  So perhaps it doesn't make sense to special case serialized attributes.

Reply:
Maybe the special casing should be for `Mutable`. I was afraid of the perf implications hence why I limited it to the cases strictly needed.

Reply:
If we need to special case serialized attributes for performance reasons, then maybe that's fine, but we should at least be explicit about that being the reason (e.g. with a comment in the code).

Reply:
```suggestion
    # Avoid causing an earlier memoized type cast of mutable serialized user values,
    # since could prevent mutations of that user value from affecting the attribute value
    # that would affect it without using this library.
    if attribute.type.is_a?(::ActiveRecord::Type::Serialized)
```